File: mobile/lib/features/suppliers/models/supplier.dart
class SupplierTransaction {
  final double amount;
  final String type;
  final String note;
  final DateTime? date;

  const SupplierTransaction({
    required this.amount,
    required this.type,
    required this.note,
    this.date,
  });

  bool get isPayment => type == 'payment';

  factory SupplierTransaction.fromJson(Map<String, dynamic> json) => SupplierTransaction(
        amount: (json['amount'] as num?)?.toDouble() ?? 0,
        type: json['type']?.toString() ?? 'charge',
        note: json['note']?.toString() ?? '',
        date: DateTime.tryParse(json['date']?.toString() ?? ''),
      );
}

class Supplier {
  final String id;
  final String name;
  final String phone;
  final String email;
  final String address;
  final String city;
  final String notes;
  final double balance;
  final List<SupplierTransaction> transactions;
  final String? shopId;
  final String? shopName;
  final DateTime? createdAt;

  const Supplier({
    required this.id,
    required this.name,
    this.phone = '',
    this.email = '',
    this.address = '',
    this.city = '',
    this.notes = '',
    this.balance = 0,
    this.transactions = const [],
    this.shopId,
    this.shopName,
    this.createdAt,
  });

  bool get hasDue => balance > 0;

  factory Supplier.fromJson(Map<String, dynamic> json) {
    final shop = json['shop'];
    String? sId;
    String? sName;
    if (shop is Map<String, dynamic>) {
      sId = (shop['id'] ?? shop['_id'])?.toString();
      sName = shop['name']?.toString();
    } else if (shop != null) {
      sId = shop.toString();
    }
    return Supplier(
      id: (json['id'] ?? json['_id']).toString(),
      name: json['name']?.toString() ?? '',
      phone: json['phone']?.toString() ?? '',
      email: json['email']?.toString() ?? '',
      address: json['address']?.toString() ?? '',
      city: json['city']?.toString() ?? '',
      notes: json['notes']?.toString() ?? '',
      balance: (json['balance'] as num?)?.toDouble() ?? 0,
      transactions: (json['transactions'] as List?)
              ?.map((e) => SupplierTransaction.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      shopId: sId,
      shopName: sName,
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? ''),
    );
  }
}

File: backend/src/models/Supplier.js
const mongoose = require('mongoose');

const balanceTransactionSchema = new mongoose.Schema(
  {
    amount: { type: Number, required: true, min: 0 },
    type: { type: String, enum: ['charge', 'payment'], required: true },
    note: { type: String, default: '' },
    date: { type: Date, default: Date.now },
    by: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
  },
  { _id: false }
);

const supplierSchema = new mongoose.Schema(
  {
    name: { type: String, required: true, trim: true },
    phone: { type: String, default: '', trim: true },
    email: { type: String, default: '', trim: true },
    address: { type: String, default: '' },
    city: { type: String, default: '' },
    notes: { type: String, default: '' },
    balance: { type: Number, default: 0, min: 0 },
    transactions: { type: [balanceTransactionSchema], default: [] },
    shop: { type: mongoose.Schema.Types.ObjectId, ref: 'Shop', required: true },
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    isDeleted: { type: Boolean, default: false },
    deletedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    deletedAt: { type: Date, default: null },
    deleteReason: { type: String, default: '' },
  },
  { timestamps: true }
);

supplierSchema.index({ shop: 1, isDeleted: 1 });
supplierSchema.index({ name: 'text', phone: 'text', email: 'text' });

module.exports = mongoose.model('Supplier', supplierSchema);
